### DISCUSSION TITLE
Promregator release as a library in addition to an executable

### DISCUSSION THREAD

**Discussion by cyberseventies** [promregator/promregator#197] [Ideas]
A project that I work on uses Promregator and has had a good experience with it so far. I'd like to add some custom functionality but since it is only released as an executable jar, I can't easily add custom functionality for our project.

I'd like to get some opinions on enabling the release as both a library and an executable so that extension and modification is simpler. The easiest way is probably to use a classifier in the spring-boot-maven-plugin, for example as described at [this link](https://www.baeldung.com/spring-boot-dependency).

There are surely pros and cons to any path, but any thoughts or ideas are appreciated. Thanks!

**Comment by checketts** [comment]
@cyberseventies Can you please list some of the custom functionality you are adding? Are you aware of any other applications that do a similar approach that would be worth mimicking?

**Reply by cyberseventies** [reply]
Hello @checketts.
First, I opened a draft PR, just so you could see what I meant.  I tried this out locally and it seems to work, but here's the PR: https://github.com/promregator/promregator/pull/198

The first bit of custom functionality is specific to the environment in which we are running our CloudFoundry instances. We do not have a set of credentials for CF api when we start instances in CF. Adding them to static config or even the commandline is impossible. We solved this by forking the promregator code for now, but I'd like to do it in a more maintainable way.

I imagine this situation is common for groups using managed, on-demand ids, but I am not sure how many are doing so. I could easily add a generic service interface and configuration for this behavior to Promregator if desired. Still, I'd need to be able to extend it locally to implement the custom code required to get the credential info on our specific platform.

I've got a few other ideas - unrelated to this - that I'd like to also suggest / contribute to if valuable, but that's for later.

Thanks!

**Reply by cyberseventies** [reply]
I was in a bit of a hurry before, but these cred systems are oftentimes referred to as "vaults" or "safes" and can be push or pull based, depending on the situation. Most will change the credentials on some schedule to a new, random value, so a fallback is needed at runtime to retrieve updated entries. 

If Promregator could be passed a service interface at startup, for example, that does the work to retrieve the username/password at runtime, I can add custom implementations that support our scenario. This is what we've done in the fork we are using currently.

Such a service interface could have a default implementation that simply takes the applied creds ( ${cf.username} and ${cf.password} ) in order to ensure compatibility for those who do not need this complexity.

Thanks.

**Reply by eaglerainbow** [reply]
> The first bit of custom functionality is specific to the environment in which we are running our CloudFoundry instances. We do not have a set of credentials for CF api when we start instances in CF. Adding them to static config or even the commandline is impossible. We solved this by forking the promregator code for now, but I'd like to do it in a more maintainable way.

Just out of curiosity: How are you setting the credentials on update? I have some ideas on how to achieve this, but I'd sense a lot of pitfalls you might get into there. Could you elaborate a bit on the general idea you have applied there (no need to share the code)?

**Reply by cyberseventies** [reply]
@eaglerainbow - I am sure you can imagine the scenario already from the code in #202, but my code implements ```CFApiCredentials``` in such a way that is able to securely retrieve, review and verify the credentials coming from our foundation. It just injects that Bean ahead of time to ensure that all calls to it are processed accordingly. Exception handling around this is, by nature, somewhat elaborate.

**Comment by eaglerainbow** [comment]
There is one more thing to consider in this game that hasn't been mentioned yet: So far, Promregator considers all its classes and interfaces internal. That is to say, Promregator does not have any "public solution model" yet, i.e. a contract of keeping interfaces stable over time. In a nutshell, being able to just have "pure Promregator" (without all its dependent libraries) is one thing - but you will then still suffer from the fact that you cannot rely on the stability of the entities you code against.

**Reply by cyberseventies** [reply]
My intent is definitely **not** to use Promregator without its dependencies.  To be clear, I'd want to be able to pull Promregator and (the reference to) all of its transitive dependencies as a single dependency to my project. The goal of my project would be to provide a custom implementation of the ```CFApiCredentials``` interface provided in #202. The custom implementation then handles all of the requirements for the vaulting / synchronization / cycling of credentials.

EDIT: To clarify: Using the library, I would implicitly want to pull all of the referenced, transitive dependencies (with the pre-defined versions, metadata, etc.) as part of the build. If a user were to change those versions via explicit overrides, it would clearly bring additional risk.  If they chose to keep the versions as defined and simply add a custom credential implementation, however, there should be no impact to the base of promregator.

**Comment by eaglerainbow** [comment]
There is another aspect to this discussion which we should consider: Many consumers seem to be fascinated/interested by the way how Promregator retrieves metadata from the CF Cloud Controller. A lot of [Promregator's issues](https://github.com/promregator/promregator/issues) rank around that, trying to "make Promregator something else that it isn't". My impression simply is that they fear the effort for implementing the retrieval logic for reading the metadata from the CFCC. 
Therefore, a while ago, we were discussing whether we couldn't split Promregator into two parts: the "application Promregator" and a library, which does all the metadata retrieval. The application internally would use the library. Yet, external tools could also refer to and reuse the library. The rocket hasn't launched yet, because @checketts and I were unhappy about the current level of complexity and the way how the retrieval logic was implemented. Efforts there, however, pattered out without big results so far (consumers are way too happy that it was worth to cause all the necessary incompatibilities).
I know that the request here is a different one (you are more talking about a custom extension than implementing "something else on top of retrieval"). Yet, I'd like that we also have this general aspect (perhaps not its first idea of implementation) in the back of our minds.

**Reply by cyberseventies** [reply]
@eaglerainbow - you make an important point here.  My goal is more just custom creds handling.  I am not in the situation of needing any split to this behavior. At the same time, I can see how folks may want to leverage the metadata retrieval. My initial reaction is that if such a split were necessary, I'd prefer to see it in such a way that:

1. Users can simply pull Promregator as an executable and use it - exactly as designed today.
2. Users can pull Promregator-library (or whatever we'd call it) and extend it, allowing them to implement custom-credentials and perhaps other custom behaviors that are far too platform-specific to be included in Promregator's base executable.
3. Some additional library ("metadata-retrieval") which could be pulled in separately for those who are interested in using only that behavior.

Note that if implemented this way, 1 and (implicitly) 2 would be able to pull 3 as a dependency.  There are lots of alternatives to all of this, though, so I'd like to have some discussion as to what would be cleanest and easiest to maintain.

**Comment by eaglerainbow** [comment]
> If Promregator could be passed a service interface at startup, for example, that does the work to retrieve the username/password at runtime, I can add custom implementations that support our scenario. 

Sounds to me more like a case of bean-based configuration, as Promregator is and will stay an application based on spring-framework. At least I would suggest to try that first, as it appears to me being the easier, more sustainable and more powerful approach.

**Reply by cyberseventies** [reply]
I'm hoping that #202 meets this.  This is the most streamlined approach I saw, but with some additional code, we could make it more flexible as well, if that is preferable.
